vortex-array/src/vtable/mod.rs

```diff
@@ -0,0 +1,116 @@
+//! This module contains the VTable definitions for a Vortex encoding.
+
+mod array;
+mod compute;
+mod decode;
+mod encode;
+mod operations;
+mod serde;
+mod validity;
+mod visitor;
+
+use std::fmt::Debug;
+use std::ops::Deref;
+
+pub use array::*;
+pub use compute::*;
+pub use decode::*;
+pub use encode::*;
+pub use operations::*;
+pub use serde::*;
+pub use validity::*;
+pub use visitor::*;
+
+use crate::{Array, Encoding, EncodingId, EncodingRef, IntoArray};
+
+/// The encoding [`VTable`] encapsulates logic for an Encoding type and associated Array type.
+/// The logic is split across several "VTable" traits to enable easier code organization than
+/// simply lumping everything into a single trait.
+///
+/// Some of these vtables are optional, such as the [`SerdeVTable`], which is only required if
+/// the encoding supports serialization.
+///
+/// From this [`VTable`] trait, we derive implementations for the sealed [`Array`] and [`Encoding`]
+/// traits via the [`crate::ArrayAdapter`] and [`crate::EncodingAdapter`] types respectively.
+///
+/// The functions defined in these vtable traits will typically document their pre- and
+/// post-conditions. The pre-conditions are validated inside the [`Array`] and [`Encoding`]
+/// implementations so do not need to be checked in the vtable implementations (for example, index
+/// out of bounds). Post-conditions are validated after invocation of the vtable function and will
+/// panic if violated.
+pub trait VTable: 'static + Sized + Send + Sync + Debug {
+    type Array: 'static + Send + Sync + Clone + Deref<Target = dyn Array> + IntoArray;
+    type Encoding: 'static + Send + Sync + Clone + Deref<Target = dyn Encoding>;
+
+    type ArrayVTable: ArrayVTable<Self>;
+    type CanonicalVTable: CanonicalVTable<Self>;
+    type OperationsVTable: OperationsVTable<Self>;
+    type ValidityVTable: ValidityVTable<Self>;
+    type VisitorVTable: VisitorVTable<Self>;
+
+    /// Optionally enable implementing dynamic compute dispatch for this encoding.
+    /// Can be disabled by assigning to the unit `()` type.
+    type ComputeVTable: ComputeVTable<Self>;
+    /// Optionally enable the [`EncodeVTable`] for this encoding. This allows it to partake in
+    /// compression.
+    /// Can be disabled by assigning to the unit `()` type.
+    type EncodeVTable: EncodeVTable<Self>;
+    /// Optionally enable serde for this encoding by implementing the [`SerdeVTable`] trait.
+    /// Can be disabled by assigning to the unit `()` type.
+    type SerdeVTable: SerdeVTable<Self>;
+
+    /// Returns the ID of the encoding.
+    fn id(encoding: &Self::Encoding) -> EncodingId;
```

Comment:
none of the existing impls seem to use this parameter, curious what it's for?

Reply:
This is sort of what I mean r.e. all vtables functions taking either an encoding or array. None of our static encodings need this parameter, because all of them have empty structs for their encoding data.

But for PyVortex (and this isn't yet implemented), we can return the ID based on whatever the user has configured: https://github.com/vortex-data/vortex/blob/1083b29452eca6305c60c9b10eb143ffd553981d/pyvortex/src/arrays/py/encoding.rs#L21

If this was purely static, then pyvortex could only use compile-time hard-coded encoding IDs, e.g. `vortex.python`, and it wouldn't be extensible.